### PR TITLE
React Navigation: headerLayoutPreset && headerBackTitleVisible

### PR DESCRIPTION
**Pull request by builtbyproxy**
https://reactnavigation.org/docs/en/stack-navigator.html

image

According to the [documentation] you can add the parameter `headerLayoutPreset` to the stackNavigator's configuration. This will ensure your title is centered in the header on android devices whereas on IOS it is centered by default. 

This PR Adds the missing types: 
- headerLayoutPreset
- headerBackTitleVisible

This PR does not include the type `transparentCard` because it is currently marked as experimental

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [X] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`

[documentation]: <https://reactnavigation.org/docs/en/stack-navigator.html>

### CODE REVIEW

**Comment by builtbyproxy** [comment]
Encountering: 
image

Any assistance/thoughts would be apprecaited as I dont think these are related to my PR

**Comment by iRoachie** [comment]
They probably aren't related to your PR but it's the same test file. You have 2 options, either fix the tests or remove them (comment them out).

**Comment by builtbyproxy** [comment]
I would rather not remove the tests so let me have a look at fixing them, in the middle of work currently but Ill get back to you on this 👍

**Comment by builtbyproxy** [comment]
Ive just pushed a fix that half addresses the problem. 
We have about 8 tests failing. 4 of them were `ref`s missing their type declaration inside an arrow function 

```
onRef={ref => {stuff}}
```

Instead of
```
onRef={((ref: someType) => { })}
```
------------------------------

The remaining four errors were due to an incorrect implementation of this [answer].
Upon further inspection these four tests were not being used anywhere. 
Additionally I do not have enough understanding of typescript to correctly fix these errors.

I have commented out the four relevant methods as you can see in this commit 
b410dda

Please let me know if this needs anything else!

Thankyou

[answer]: <https://stackoverflow.com/questions/31815633/what-does-the-error-jsx-element-type-does-not-have-any-construct-or-call>

**Review by iRoachie — APPROVED** [review]

**Comment by iRoachie** [comment]
Thank you man! This is awesome work 💯

File: types/react-navigation/react-navigation-tests.tsx [hunk 1]
```diff
@@ -552,32 +557,37 @@ class MyBackButton extends React.Component<BackButtonProps & NavigationInjectedP
...
-const BackButtonWithNavigation = withNavigation(MyBackButton);
-const BackButtonInstance = <BackButtonWithNavigation
-    title="Back" onRef={ref => {
-        // ref is inferred as MyBackButton | null
-        if (!ref) return;
-        ref.triggerBack();
-    }}
-/>;
+
+// This is not used and is causing an error with the assignment to BackButtonWithNavigation below
```
Comment (sheetalkamat) [inline types/react-navigation/react-navigation-tests.tsx]: I dont think I understand what this comment means. I would like other reviewers to comment on this before merging. Thanks.

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/react-navigation/react-navigation-tests.tsx (hunk 1 above)
Comment (builtbyproxy) [reply types/react-navigation/react-navigation-tests.tsx]: Hey Sheetal

The four implementations of this comment all reference an instance of a component instead of a constructor. This then causes the error: 
`
JSX element type <ElementInQuestion> does not have any construct or call signatures
`

Please see [this] for a detailed explanation.

Additionally, according to my VSCode the code that is causing these errors is not actually used anywhere anyway.

[this]: <https://stackoverflow.com/questions/31815633/what-does-the-error-jsx-element-type-does-not-have-any-construct-or-call>

Comment (builtbyproxy) [reply types/react-navigation/react-navigation-tests.tsx]: I dont like the idea of disabling these even if they arent being used however my knowledge of typescript is super limited for this problem... Any insight would be greatly appreciated!

**Comment by builtbyproxy** [comment]
@sheetalkamat Hey Sheetal would you mind addressing the conversation above whenever you have some time free please? Thankyou.

**Comment by builtbyproxy** [comment]
@iRoachie 	Could you please re-open this PR as I dont believe it should be closed

**Comment by iRoachie** [comment]
@builtbyproxy I'm actually not in control of that.

**Comment by builtbyproxy** [comment]
> @builtbyproxy I'm actually not in control of that.

Ah all good, thankyou for getting back to me though

**Comment by builtbyproxy** [comment]
@sheetalkamat When you are free could you please offer some more input on this? 
Thankyou

~Lachlan

**Comment by builtbyproxy** [comment]
@andy-ms 	
Sorry to drag you into this, could you possibly help with this? I see you are one of the top contributors.

**Comment by andy-ms** [comment]
@builtbyproxy Please revert the commented-out code and we'll see if we can get this fixed again now that Microsoft/TypeScript#27948 was marked resolved.

**Comment by builtbyproxy** [comment]
@sheetalkamat 	Any update on this for you, it is passing all current tests now.

File: types/react-navigation/react-navigation-tests.tsx [hunk 2]
```diff
@@ -553,31 +558,35 @@ class MyBackButton extends React.Component<BackButtonProps & NavigationInjectedP
...
-    }}
+    })}
 />;
 
 function StatelessBackButton(props: BackButtonProps & NavigationInjectedProps) {
   return <MyBackButton {...props} />;
 }
 
 // Wrapped stateless components don't accept an onRef
+// This is not used and is causing an error with the assignment to StatelessBackButtonWithNavigation below
```
Comment (andy-ms) [inline types/react-navigation/react-navigation-tests.tsx]: Comments should be removed

File: package.json [hunk 3]
```diff
@@ -23,5 +23,6 @@
     "devDependencies": {
         "dtslint": "github:Microsoft/dtslint#production",
         "types-publisher": "Microsoft/types-publisher#production"
-    }
+    },
+    "dependencies": {}
```
Comment (andy-ms) [inline package.json]: Please undo this change -- don't see why this is necessary

**Comment by andy-ms** [comment]
This PR just adds two properties, right? None of the tests changes seem relevant.

**Comment by builtbyproxy** [comment]
Agreed, they aren't relevant hence undoing them. However they were failing before I merged master after resolving [Microsoft/TypeScript#27948](https://github.com/Microsoft/TypeScript/issues/27948)

**Comment by andy-ms** [comment]
> I believe the addition of headerLayoutPreset to the tests along with defining the types of the refs is beneficial though. Thoughts?

Sure. But the other changes should go in a different PR if at all.

**Comment by builtbyproxy** [comment]
Are you talking about defining the types of the refs? Would you mind flagging the lines of concern please?

**Comment by andy-ms** [comment]
Everything besides `const headerLayoutPreset = 'center';` and its use.

**Comment by builtbyproxy** [comment]
Oh!
I cant even keep headerBackTitleVisible?

File: types/react-navigation/react-navigation-tests.tsx [hunk 4]
```diff
@@ -47,6 +47,7 @@ import {
     withNavigationFocus,
     NavigationFocusInjectedProps
 } from 'react-navigation';
+import { string } from 'prop-types';
```
Comment (builtbyproxy) [inline types/react-navigation/react-navigation-tests.tsx]: So would you like me to revert this

File: types/react-navigation/react-navigation-tests.tsx [hunk 5]
```diff
@@ -125,6 +126,9 @@ const initialRouteParams: StartScreenNavigationParams = {
     id: 1,
     s: "Start",
 };
+
+const headerLayoutPreset = 'center';
```
Comment (builtbyproxy) [inline types/react-navigation/react-navigation-tests.tsx]: and this

File: types/react-navigation/react-navigation-tests.tsx [hunk 6]
```diff
@@ -140,6 +144,7 @@ export const AppNavigator = createStackNavigator(
     {
         initialRouteName: ROUTE_NAME_START_SCREEN,
         initialRouteKey: ROUTE_KEY_START_SCREEN,
+        headerLayoutPreset,
```
Comment (builtbyproxy) [inline types/react-navigation/react-navigation-tests.tsx]: and this

File: types/react-navigation/react-navigation-tests.tsx [hunk 7]
```diff
@@ -555,11 +560,11 @@ class MyBackButton extends React.Component<BackButtonProps & NavigationInjectedP
 // If you have class methods, you should have a way to use them.
 const BackButtonWithNavigation = withNavigation(MyBackButton);
 const BackButtonInstance = <BackButtonWithNavigation
-    title="Back" onRef={ref => {
+    title="Back" onRef={((ref: MyBackButton) => {
```
Comment (builtbyproxy) [inline types/react-navigation/react-navigation-tests.tsx]: and this

File: types/react-navigation/react-navigation-tests.tsx [hunk 8]
```diff
@@ -573,11 +578,11 @@ const StatelessBackButtonInstance = <StatelessBackButtonWithNavigation title="Ba
 // The old way of passing in the props should still work
 const BackButtonWithNavigationWithExplicitProps = withNavigation<BackButtonProps>(MyBackButton);
 const BackButtonWithExplicitPropsInstance = <BackButtonWithNavigationWithExplicitProps
-    title="Back" onRef={ref => {
+    title="Back" onRef={((ref: React.Component<BackButtonProps & NavigationInjectedProps>) => {
```
Comment (builtbyproxy) [inline types/react-navigation/react-navigation-tests.tsx]: and this

File: types/react-navigation/react-navigation-tests.tsx [hunk 9]
```diff
@@ -593,7 +598,7 @@ class MyFocusedComponent extends React.Component<MyFocusedComponentProps & Navig
 // navigation and isFocused prop
 const MyFocusedComponentWithNavigationFocus = withNavigationFocus(MyFocusedComponent);
 const MyFocusedComponentInstance = <MyFocusedComponentWithNavigationFocus
-    expectsFocus={true} onRef={ref => { const backButtonRef = ref; }}
+    expectsFocus={true} onRef={((ref: MyFocusedComponent) => { const backButtonRef = ref; })}
```
Comment (builtbyproxy) [inline types/react-navigation/react-navigation-tests.tsx]: and this

**Comment by builtbyproxy** [comment]
haha @andy-ms 	Im sure you're busy so ive just commented on all test changes, could you let me know what is ok and Ill revert the ones you say no to :)

File: types/react-navigation/react-navigation-tests.tsx (hunk 9 above)
Comment (andy-ms) [reply types/react-navigation/react-navigation-tests.tsx]: Y

File: types/react-navigation/react-navigation-tests.tsx (hunk 8 above)
Comment (andy-ms) [reply types/react-navigation/react-navigation-tests.tsx]: Y

File: types/react-navigation/react-navigation-tests.tsx (hunk 7 above)
Comment (andy-ms) [reply types/react-navigation/react-navigation-tests.tsx]: Y

File: types/react-navigation/react-navigation-tests.tsx (hunk 6 above)
Comment (andy-ms) [reply types/react-navigation/react-navigation-tests.tsx]: N (undo the shorthand and use `headerLayoutPreset: 'center',`)

File: types/react-navigation/react-navigation-tests.tsx (hunk 5 above)
Comment (andy-ms) [reply types/react-navigation/react-navigation-tests.tsx]: Y

File: types/react-navigation/react-navigation-tests.tsx (hunk 4 above)
Comment (andy-ms) [reply types/react-navigation/react-navigation-tests.tsx]: Y

**Comment by builtbyproxy** [comment]
@andy-ms thanks for that! How does it look now

**Comment by builtbyproxy** [comment]
Thanks for all your help
